contracts/modules/STO/POLYCappedSTO.sol

```diff
@@ -0,0 +1,526 @@
+    ) internal {
+        if (!allowBeneficialInvestments) {
+            require(_beneficiary == msg.sender, "Beneficiary != msg.sender");
+        }
+        require(_canBuy(_beneficiary), "Unauthorized");
+        uint256 _spentValue = _buyTokens(_beneficiary, _tokenAmount, _fundRaiseType);
+        // Forward coins to issuer wallet
+        require(_token.transferFrom(msg.sender, wallet, _spentValue), "Transfer failed");
+        emit FundsReceived(msg.sender, _beneficiary, _fundRaiseType, _tokenAmount, _spentValue);
+        _postValidatePurchase(_beneficiary, _spentValue);
+    }
+
+    /**
+      * @notice Low level token purchase
+      * @param _beneficiary Address where security tokens will be sent
+      * @param _investmentValue Amount of POLY invested
+      * @param _fundRaiseType Fund raise type (POLY)
+      */
+    function _buyTokens(
+        address _beneficiary,
+        uint256 _investmentValue,
+        FundRaiseType _fundRaiseType
+    )
+        internal
+        nonReentrant
+        whenNotPaused
+        returns(uint256 _spentValue)
+    {
+        uint256 _tokens;
+        (_tokens, _spentValue) = prePurchaseChecks (_beneficiary, _investmentValue);
+        _processPurchase(_beneficiary, _tokens);
+        uint256 polyUsdRate = IOracle(PolymathRegistry(RegistryUpdater(securityToken).polymathRegistry()).getAddress(POLY_ORACLE)).getPrice();
+        uint256 _spentUSD = DecimalMath.mul(_spentValue, polyUsdRate);
+        emit TokenPurchase(msg.sender, _beneficiary, _spentValue, _spentUSD, _tokens);
+        _updatePurchasingState(_beneficiary, _investmentValue);
+        // Modify storage
+        investorInvested[_beneficiary] = investorInvested[_beneficiary].add(_spentValue);
+        fundsRaised[uint8(_fundRaiseType)] = fundsRaised[uint8(_fundRaiseType)].add(_spentValue);
+        investorInvestedUSD[_beneficiary] = investorInvestedUSD[_beneficiary].add(_spentUSD);
+        fundsRaisedUSD = fundsRaisedUSD.add(_spentUSD);
+        totalTokensSold = totalTokensSold.add(_tokens);
+    }
+
+    /**
+      * @notice Checks restrictions related to the token purchase and calculates number of tokens
+      * @notice and calculates the number of tokens and spent value based on those restrictions
+      * @param _beneficiary Address where security tokens will be sent
+      * @param _investmentValue Amount of POLY invested
+      * @return _tokens Number of tokens the _beneficiary will recieve
+      * @return _spentValue Number of POLY that will be spent
+      */
+    function prePurchaseChecks(address _beneficiary, uint256 _investmentValue) public view returns(
+        uint256 _tokens,
+        uint256 _spentValue
+    ) {
+        //Pre-Purchase checks
+        require(isOpen(), "STO not open");
+        require(_investmentValue > 0, "No funds were sent");
+        require(_beneficiary != address(0), "Beneficiary address should not be 0x");
+        if (!_isAccredited(_beneficiary) && maxNonAccreditedInvestors != 0) {
```

Comment:
'!_isAccredited(_beneficiary)` is being checked at a lot of places in a single purchase call. This can be optimized to do all actions regarding Accredited in a single place.

Reply:
I've refactored this so _isAccredited() is only called once and consolidated the non accredited pre-purchase checks. The pre-purchase check now returns the accredited status so the non accredited investor count can be updated only when a purchase is being made. I did it this way as the `prePurchaseCheck` function is public allowing an investor to verify exactly how many tokens they will get and how much it will cost before a purchase is made.